core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -350,7 +372,8 @@ class BrokerToControllerRequestThread(
     } else if (response.wasDisconnected()) {
       updateControllerAddress(null)
       requestQueue.putFirst(queueItem)
-    } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER)) {
+    } else if (response.responseBody().errorCounts().containsKey(Errors.NOT_CONTROLLER) ||
+      maybeCheckNotControllerErrorInsideEnvelopeResponse(queueItem.requestHeader, response.responseBody())) {
```

Comment:
@showuon Because it's all checking for Not Controller error, can we put the logic for check NotControllerError into the same method? So the else if argument is a little bit cleaner. But this is only a small tip, and it's up to you to decide if you need to do it. This is a great discovery. You have solved my confusion when I met UNKNOWN_SERVER_ERROR. Thank you.

Reply:
@wenbingshen , thanks for the comment. I've addressed your comment. Thank you.